### PR TITLE
Dev team chat

### PR DESCRIPTION
**Pull request by S34NW**
## What Does This PR Do
Adds an ingame chat for dev team members to communicate with each other and admins.
Adds a dev team permission to access this chat.
Refactors who code to be less insane, adds a dev section to it.

## Why It's Good For The Game
Better cross-team and inter-team communications is good.

## Images of changes
delwedd
delwedd

delwedd

## Testing
Tried to speak as admin, dev member, and non-admin. Everything worked as intended.

<hr>

### Declaration

- [x] I confirm that I either do not require [pre-approval](https://github.com/ParadiseSS13/Paradise/blob/master/CODE_OF_CONDUCT.md#types-of-changes-that-need-approval) for this PR, or I have obtained such approval and have included a screenshot to demonstrate this below.
  
  
  <hr>

## Changelog

:cl:
add: Added developer team chat
/:cl:

### CODE REVIEW

File: code/modules/admin/verbs/adminsay.dm [hunk 1]
```diff
@@ -45,6 +45,48 @@
...
+	if(!msg)
+		return
+
+	// Do this up here before it gets sent to everyone & emoji'd
+	if(SSredis.connected)
+		var/list/data = list()
+		data["author"] = usr.ckey
+		data["source"] = GLOB.configuration.system.instance_id
+		data["message"] = html_decode(msg)
+		SSredis.publish("byond.devsay", json_encode(data))
```
Comment (AffectedArc07) [inline code/modules/admin/verbs/adminsay.dm:75]: Todo - Alice update.

Dunno if we need #devsay, probably to not bridge it all to the main chat

File: code/modules/admin/verbs/adminsay.dm [hunk 2]
```diff
@@ -45,6 +45,48 @@
...
+			var/display_name = key
+			if(holder.fakekey)
+				if(C.holder && C.holder.rights & R_ADMIN)
+					display_name = "[holder.fakekey]/([key])"
+				else
+					display_name = holder.fakekey
+			msg = "<span class='emoji_enabled'>[msg]</span>"
+			to_chat(C, "<span class='[check_rights(R_ADMIN, 0) ? "dev_channel_admin" : "dev_channel"]'>DEV: <span class='name'>[display_name]</span> ([admin_jump_link(mob)]): <span class='message'>[msg]</span></span>", MESSAGE_TYPE_DEVCHAT, confidential = TRUE)
+
+	SSblackbox.record_feedback("tally", "admin_verb", 1, "Msay") //If you are copy-pasting this, ensure the 2nd parameter is unique to the new proc!
```
Comment (AffectedArc07) [inline code/modules/admin/verbs/adminsay.dm]: See comment

File: config/example/config.toml [hunk 3]
```diff
@@ -116,7 +120,7 @@ admin_rank_colour_map = [
 	{ name = "Community Manager", colour = "#e91e63" },
 	{ name = "Game Admin", colour = "#238afa" },
 	{ name = "Trial Admin", colour = "#7fb6fc" },
-	{ name = "PR Reviewer", colour = "#c27c0e" },
+	{ name = "Developer", colour = "#c27c0e" },
```
Comment (AffectedArc07) [inline config/example/config.toml:123]: Is it worth migrating people with +VIEWRUNTIMES to this role?

**Review by AffectedArc07 — COMMENTED** [review]
Thought a fair bit on this, is it worth just converting +VIEWRUNTIMES into +DEV?

Trying to think of an instance where we need +RUNTIMES but not access to a private chat to discuss current TM.

File: config/example/config.toml (hunk 3 above)
Comment (S34NW) [reply config/example/config.toml:123]: We sometimes give viewruntimes to people not on the dev team, although it is rare

**Comment by ROdenFL** [comment]
kinda funny. two weeks ago made almost similiar thing https://github.com/ss220-space/Paradise/pull/5765

File: tgui/packages/tgui-panel/styles/tgchat/chat-dark.scss [hunk 4]
```diff
@@ -405,6 +405,16 @@ em {
   font-weight: bold;
 }
 
+.dev_channel {
+  color: #775bff;
```
Comment (Contrabang) [inline tgui/packages/tgui-panel/styles/tgchat/chat-dark.scss:409]: These colors are really similar to deadchat colors from what I can see, maybe it should have a unique color. Maybe something green or pink

File: config/example/config.toml (hunk 3 above)
Comment (Burzah) [inline config/example/config.toml:123]: Can we change the color of this to what the dev say color (green color) is? It already looks to similar to patreon colors.

How will colors and permissions behave if you are both a mentor and a developer?

File: tgui/packages/tgui-panel/styles/tgchat/chat-dark.scss (hunk 4 above)
Comment (S34NW) [reply tgui/packages/tgui-panel/styles/tgchat/chat-dark.scss:409]: They are the same as the mentor channel, might be worth a separate re-evaluation of our chat colours for these things

File: config/example/config.toml (hunk 3 above)
Comment (S34NW) [reply config/example/config.toml:123]: Dev takes priority over mentor

Comment (S34NW) [reply config/example/config.toml:123]: Also yes let me change that colour

File: code/game/verbs/who.dm [hunk 5]
```diff
@@ -3,46 +3,44 @@
...
 	for(var/client/C in GLOB.clients)
 		if(C.holder && C.holder.big_brother) // BB doesn't show up at all
 			continue
 
 		if(C.holder && C.holder.fakekey)
-			temp += C.holder.fakekey
+			ckeys.Add(C.holder.fakekey)
 		else
-			temp += C.key
+			ckeys.Add(C.key)
```
Comment (AffectedArc07) [inline code/game/verbs/who.dm:15]: iirc += is faster

(to be clear I hate that that is the fact)

**Review by AffectedArc07 — COMMENTED** [review]
Up to you if you want to change `.Add()` to `+=`.

Im fine to merge this as is

File: code/game/verbs/who.dm (hunk 5 above)
Comment (S34NW) [reply code/game/verbs/who.dm:15]: We standardised the other way I think? Cc @ParadiseSS13/headcoders

File: code/__DEFINES/admin_defines.dm [hunk 6]
```diff
@@ -40,11 +39,12 @@
...
 #define R_VIEWRUNTIMES	(1<<16)
 #define R_MAINTAINER	(1<<17)
+#define R_DEV_TEAM		(1<<18)
 // Update the following two defines if you add more
 
-#define R_MAXPERMISSION (1<<17) //This holds the maximum value for a permission. It is used in iteration, so keep it updated.
+#define R_MAXPERMISSION (1<<18) // This holds the maximum value for a permission. It is used in iteration, so keep it updated.
 
-#define R_HOST			((1<<18)-1) //17 bit bitmask, update me if we ever add more admin permissions. Sum of all permissions to allow easy setting.
+#define R_HOST			((1<<19)-1) // 18 bit bitmask, update me if we ever add more admin permissions. Sum of all permissions to allow easy setting.
```
Comment (ROdenFL) [inline code/__DEFINES/admin_defines.dm:47]:
```suggestion
#define R_HOST			(~0) // Sum of all permissions to allow easy setting.
```
i don`t know, good this solution or not, but it works

Comment (S34NW) [reply code/__DEFINES/admin_defines.dm:47]: delwedd
Seems legit

File: config/example/config.toml [hunk 7]
```diff
@@ -64,6 +64,7 @@ enable_localhost_autoadmin = true
 # +SPAWN (or +CREATE) = mob transformations, spawning of most atoms including mobs
 # +MOD = Access to low level moderation tools. Not used in Paradise production
 # +MENTOR = Access only to the Question's Ahelp and has little way of metagaming the game.
+# +DEV_TEAM = Access to dev team chat
 # +PROCCALL = Access to call procs on anything. Be careful who you grant this too!
 # +VIEWRUNTIMES = Allows a player to view the runtimes of the server, but not use other debug verbs
```
Comment (AffectedArc07) [inline config/example/config.toml:69]:
```suggestion
# +PROCCALL = Access to call procs on anything. Be careful who you grant this too!
# +VIEWRUNTIMES = Allows a player to view the runtimes of the server, but not use other debug verbs
# +MAINTAINER = Access to super dangerous debug verbs that can break an entire round. Use sparingly.
# +DEV_TEAM = Access to dev team chat
```
Smallest nit ever, just keep it inline with the game order

File: config/example/config.toml [hunk 8]
```diff
@@ -116,7 +120,7 @@ admin_rank_colour_map = [
 	{ name = "Community Manager", colour = "#e91e63" },
 	{ name = "Game Admin", colour = "#238afa" },
 	{ name = "Trial Admin", colour = "#7fb6fc" },
-	{ name = "PR Reviewer", colour = "#c27c0e" },
+	{ name = "Developer", colour = "#61b413" },
```
Comment (AffectedArc07) [inline config/example/config.toml:123]: image

**Review by AffectedArc07 — APPROVED** [review]
Ready after that one config nit

**Pull request merged**